CIPs/cip-0032.md

```diff
@@ -0,0 +1,117 @@
+
+If we observe that a validator is relying too heavily upon the `SignalOnline` functionality when they're receiving legitimate attestation requests we can add mitigations such as introducing an increasing financial penalty for subsquent signals or by decreasing the time that signal will last for (lowering `S`) per user.
+
+### Requiring Attestation Service
+
+**Note that the slasher expects each validator to be running an Attestation Service.** Unlike today, where it's optional for a Validator to run an attestation node, this proposal would make it mandatory. Having more attestation nodes in the pool improves the security of the phone number verification by making it more difficult to game the random selection of a set of colluding issuers. This proposal recommends providing at least three weeks from time of merge of this CIP to the time of enacting the slashing mechanism to give everyone an opportunity to set up their Attestation Service.
+
+### Nothing at stake problem
+
+Unlike other slashing contracts which use `SlasherUtil.performSlashing()` to both slash _and_ remove the validator, this `AttestationSlasher` would directly call `LockedGold.slash()`, only slashing stake and not removing the validator. This decreases the potential risk of having multiple validators removed simultaneously due to an overly aggressive `AttestationSlasher` policy. If the validator is slashed it may be that their locked stake drops below the `validatorLockedGoldRequirements` (currently 10K CELO). By dropping below `validatorLockedGoldRequirements` alone, the validator is not removed from the validator set but they will cease to earn epoch rewards. If they continue to get slashed by the `AttestationSlasher`, it may be that they drop to a near-zero stake while still remaining to be a validator. This is a dangerous situation, since this validator is performing an important role for the network but has nothing at stake. To resolve this tension, we'll introduce a new `MinStakeSlasher` which will force a validator to be de-affiliated if their balance drops below the `penalty + reward` threshold of any of the registered slasher contracts. Once a validator has been de-affiliated, they'll need to return their stake amount to the `validatorLockedGoldRequirements` requirement before being able to affiliate again.
+
+### Defining parameters
+
+#### AttestationSlasher
+
+- `B` This will be chosen based on the minimal period that it is likely for a validator to have recieved an attestation request. We'll start conservative and look at a period where every validator should have received 10 attestations. In the month of November 2020, there was an average of 94 attestations requested per day (more in December). Therefore, over 11 days, an average of 1034 attesations would be requested, providing an expected value of ~10 attestations for each of the 100 issuers. Since we produce a block every 5 seconds, 11 days is equivalent to 190,080 blocks.
```

Comment:
Curious, with this parameter at the suggested value, would any validators have been slashable in the past few months?

11 days to be down if very long...but I understand due to low traffic, services get fewer chances to perform. makes sense to start conservative and update when we have more traffic

Reply:
Could this value be dynamic and provided by an oracle based on the situation in the current period?

Reply:
Looking at the 11 days from 1/1 to 1/11 inclusive, there are two such validators that are registered that have 100% failure rates (0xeD32008dB382d078bABB0eB6F91Df0207d9098990, x3BEbD7Ed4cA0ce7BFd69fC6aeF31dDD21C460158)
Here's the query for future:
```
SELECT DISTINCT issuer
FROM `celo-testnet-production.analytics_eksportisto.attestation_state_with_user_data`
WHERE timestamp >= "2021-01-01"
  AND timestamp <= "2021-01-11"
AND issuer NOT IN
(SELECT DISTINCT issuer FROM `celo-testnet-production.analytics_eksportisto.attestation_state_with_user_data`
  WHERE timestamp >= "2021-01-01"
  AND timestamp <= "2021-01-11"
  AND isCompleted)
  ```
  This doesn't include the four validators which don't have attestation services registered.

Reply:
> Could this value be dynamic and provided by an oracle based on the situation in the current period?
I think to start with, it'd be safest to use governance proposals to adjust the parameters, but I can totally imagine this threshold being a function of the total number of attestations over the past week/month.

Reply:
Ya, makes sense to start simple and conservative. Can always tweek the static value or make it dynamic later.